sql/core/src/main/scala/org/apache/spark/sql/execution/SparkScriptTransformationExec.scala

```diff
@@ -0,0 +1,187 @@
+
+  override def processIterator(inputIterator: Iterator[InternalRow], hadoopConf: Configuration)
+  : Iterator[InternalRow] = {
+    val cmd = List("/bin/bash", "-c", script)
+    val builder = new ProcessBuilder(cmd.asJava)
+
+    val proc = builder.start()
+    val inputStream = proc.getInputStream
+    val outputStream = proc.getOutputStream
+    val errorStream = proc.getErrorStream
+
+    // In order to avoid deadlocks, we need to consume the error output of the child process.
+    // To avoid issues caused by large error output, we use a circular buffer to limit the amount
+    // of error output that we retain. See SPARK-7862 for more discussion of the deadlock / hang
+    // that motivates this.
+    val stderrBuffer = new CircularBuffer(2048)
+    new RedirectThread(
+      errorStream,
+      stderrBuffer,
+      "Thread-ScriptTransformation-STDERR-Consumer").start()
+
+    val finalInput = input.map(Cast(_, StringType).withTimeZone(conf.sessionLocalTimeZone))
+
+    val outputProjection = new InterpretedProjection(finalInput, child.output)
+
+    // This new thread will consume the ScriptTransformation's input rows and write them to the
+    // external process. That process's output will be read by this current thread.
+    val writerThread = ScriptTransformationWriterThread(
+      inputIterator.map(outputProjection),
+      input.map(_.dataType),
+      ioschema,
+      outputStream,
+      proc,
+      stderrBuffer,
+      TaskContext.get(),
+      hadoopConf
+    )
+
+    val reader = new BufferedReader(new InputStreamReader(inputStream, StandardCharsets.UTF_8))
+    val outputIterator: Iterator[InternalRow] = new Iterator[InternalRow] {
+      var curLine: String = null
+      val mutableRow = new SpecificInternalRow(output.map(_.dataType))
+
+      override def hasNext: Boolean = {
+        try {
+          if (curLine == null) {
+            curLine = reader.readLine()
+            if (curLine == null) {
+              checkFailureAndPropagate(writerThread, null, proc, stderrBuffer)
+              return false
+            }
+          }
+          true
+        } catch {
+          case NonFatal(e) =>
+            // If this exception is due to abrupt / unclean termination of `proc`,
+            // then detect it and propagate a better exception message for end users
+            checkFailureAndPropagate(writerThread, e, proc, stderrBuffer)
+
+            throw e
+        }
+      }
+
+      override def next(): InternalRow = {
+        if (!hasNext) {
+          throw new NoSuchElementException
+        }
+        val prevLine = curLine
+        curLine = reader.readLine()
+        if (!ioschema.schemaLess) {
+          new GenericInternalRow(
+            prevLine.split(ioschema.outputRowFormatMap("TOK_TABLEROWFORMATFIELD"))
+              .zip(output)
+              .map { case (data, dataType) =>
+                CatalystTypeConverters.convertToCatalyst(wrapper(data, dataType.dataType))
```

Comment:
> To avoid the patten-matching of data types per row, could you initialize the converters for a given schema first, then use it? e.g.,
> https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/InterpretedUnsafeProjection.scala#L136-L244

I have tried these code yesterday, but since here we got is just string m, directly use InterpretedUnsafeProject still have problem of type convert.

Reply:
You meant we cannot write it like this?
```
      private lazy val fieldWriters: Seq[String => Any] = output.map { attr =>
        val converter = CatalystTypeConverters.createToCatalystConverter(attr.dataType)
        attr.dataType match {
          case StringType => (data: String) => converter(data)
          case IntegerType => (data: String) => converter(data.toInt)
          // more types here...
          case _ => data: String => converter(data)
        }
      }

      override def next(): InternalRow = {
        if (!hasNext) {
          throw new NoSuchElementException
        }
        if (outputSerde == null) {
          val prevLine = curLine
          curLine = reader.readLine()
          if (!ioschema.schemaLess) {
            new GenericInternalRow(
              prevLine.split(ioschema.outputRowFormatMap("TOK_TABLEROWFORMATFIELD"))
                .zip(fieldWriters)
                .map { case (data, writer) => writer(data) })
          } else {
            new GenericInternalRow(
              prevLine.split(ioschema.outputRowFormatMap("TOK_TABLEROWFORMATFIELD"), 2)
                .zip(fieldWriters)
                .map { case (data, writer) => writer(data) })
          }
        } else {
```

Reply:
@maropu Got your point , pass function to avoid unnecessary match ....my poor code skills...
Updated!

Reply:
Thanks!